platforms/src/Civic/Providers/civic.ts

```diff
@@ -0,0 +1,136 @@
+    chain: SupportedChain
+  ): Promise<PassResponse | null>[] => {
+    const { rpcUrl } = EVM_CHAIN_CONFIG[chain];
+    const ethersProvider: StaticJsonRpcProvider = new StaticJsonRpcProvider(rpcUrl);
+    const passResponsesForChain = passTypes.map(getPassResponseForChain(chain, ethersProvider));
+    return [...passResponses, ...passResponsesForChain];
+  };
+
+  // iterate over all supported chains and check for passes
+  const responses = await Promise.all(chains.reduce(passReducer, []));
+
+  // remove nulls (representing "no pass of type x on chain y") from the array
+  return responses.filter((response) => response !== null);
+}
+
+type CivicPassProviderOptions = {
+  chains?: SupportedChain[];
+  passTypes?: CivicPassType[];
+  type?: string;
+  includeTestnets?: boolean;
+};
+
+// Export a Civic Pass Provider
+export class CivicPassProvider implements Provider {
+  // Give the provider a type so that we can select it with a payload
+  type = "Civic";
+
+  // Options can be set here and/or via the constructor
+  _options = {
+    chains: Object.keys(EVM_CHAIN_CONFIG) as SupportedChain[],
+    passTypes: supportedCivicPassTypes,
+    type: "CivicAllPasses",
+    includeTestnets: false,
+  };
+
+  // construct the provider instance with supplied options
+  constructor(options: CivicPassProviderOptions = {}) {
+    this._options = { ...this._options, ...options };
+    this.type = options.type;
+    // filter unsupported chains or non-mainnet chains (if includeTestnets is false)
+    this._options.chains = this._options.chains.filter(
+      (chain) => EVM_CHAIN_CONFIG[chain] && (this._options.includeTestnets || EVM_CHAIN_CONFIG[chain].mainnet)
+    );
+  }
+
+  // Verify that address defined in the payload has a civic pass
+  async verify(payload: RequestPayload): Promise<VerifiedPayload> {
+    // if a signer is provider we will use that address to verify against
+    const address = payload.address.toString().toLowerCase();
+    const passResponses = await findAllPasses(address, this._options.chains, this._options.passTypes);
+
+    const error = passResponses
+      .map((response) => response.error?.error)
+      .filter((errorMessage) => errorMessage !== undefined);
+    const passes = passResponses.map((response) => response.pass).filter((pass) => pass !== undefined);
+    const valid = passes.length > 0;
+
+    // convert the passes found to a record data structure of the form:
+    // "passType": "chain1,chain2"
+    // e.g. "UNIQUENESS": "ETHEREUM_MAINNET,POLYGON_POS_MAINNET"
+    const recordData = passes.reduce((acc, pass) => {
+      const passData = acc[CivicPassType[pass.type]] || "";
+
+      return {
+        ...acc,
+        [CivicPassType[pass.type]]: `${passData}${passData ? "," : ""}${pass.chain}`,
+      };
+    }, {} as Record<string, string>);
+
+    return {
+      valid,
+      error,
+      record: valid
+        ? {
+            address: address,
+            ...recordData,
```

Comment:
The "record" value is hashed in order to create the stamp hash which prevents users from using the same credential (civic pass) with multiple passports. I worry that including this recordData will allow malicious users to e.g. use their civic pass with one passport, then go add it to one more chain, then use their same civic pass (which will have a different hash value) with a second passport. Or, if you guys add support for a new chain and users add this, it would change their stamp hash value. Does that make sense? Do you agree, or am I missing something? I'm thinking we should just use `{address}` for the record value. Thanks!!

Reply:
Hi @lucianHymer , the passes are associated with a given wallet, so if the user tries to create a second passport they shouldn't be able to use the same passes. Does that make sense? I also might be missing something :)

The reasoning for adding the record data to the credential is that the passes come in various flavours (liveness, uniqueness, full KYC etc), and can be issued on multiple chains. Depending on the types of passes, and the chains they have them on, that might influence the scoring. For example, a uniqueness pass on ethereum would probably be worth more than a liveness pass on mumbai. I didn't want to lose that information.

I have mitigated this to some extent by making the different pass flavours different credentials, but the chain information is still retained inside the credentials themselves. In addition, credentials have metadata such as expiry which may be valuable for scoring.